Use live region in error text input decorator for Android

ash2moon opened the pull request:
Resolves partly https://github.com/flutter/flutter/issues/165510


**Context:** This issue originates from https://github.com/flutter/flutter/issues/99715, where it was reported that `liveRegion` alone was insufficient for announcing form validation errors. While `liveRegion` announces the first error encountered, subsequent submissions with the same error message on Android would not trigger a re-announcement.

**Original Solution:** Pull request https://github.com/flutter/flutter/pull/123373 addressed this by implementing the `announce` event to ensure error messages were consistently announced, even for repeated submissions.

**Native Android Behavior (Jetpack Compose):** In native Android development using Jetpack Compose, setting the `isError` property of a `TextField` to `true` triggers Talkback to announce "Error invalid input."  This announcement occurs *only* on the initial change to the error state. Subsequent errors, even if the `isError` property remains `true`, are not re-announced. This behavior closely mirrors the functionality of `liveRegion`, with the key difference being that `liveRegion` also announces the specific error text, in addition to the general error state.  Testing in a native Jetpack Compose application confirms this behavior and provides a valuable comparison point against the current Flutter form example.

**Suggested Action:** **Fork** the behavior in https://github.com/flutter/flutter/pull/123373. Reinstate the use of `liveRegion` for error announcements within `widgets/Form` for Android and keep other platforms the same.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

ash2moon committed:
remove usages of semantic announcement in favor of liveRegion due to deprecation of announcement in Android

# Conflicts:
#	packages/flutter/test/material/text_field_test.dart

hannah-hyj commented:
The original issue  https://github.com/flutter/flutter/issues/99715 is on macos with voice over, can you test if the lice region works there?

ash2moon commented:
> The original issue #99715 is not on android but on macOS with voiceover, can you test if the live region works there?

Looks like macos has issues with liveregion and not speaking this text. Does macos not support liveregion at all?

ash2moon committed:
remove usages of semantic announcement in favor of liveRegion due to deprecation of announcement in Android

ash2moon commented:
> The original issue #99715 is not on android but on macOS with voiceover, can you test if the live region works there?

I added back the original code for the other platforms and turned on liveRegion and removed the deprecated announce call for Android.

hannah-hyj commented on packages/flutter/lib/src/widgets/form.dart:
```diff
@@ -369,7 +369,7 @@ class FormState extends State<Form> {
     }
 
-    if (errorMessage.isNotEmpty) {
+    if (errorMessage.isNotEmpty && defaultTargetPlatform != TargetPlatform.android) {
```
can you add a comment here that android is use live region for error message and others are using announcement

reidbaker replied:
Can you instead refactor this into a method named "shouldPreferLiveRegionOverAnnounce"

ash2moon replied:
I ended up choosing to go with `isAnnounceSupported` because live region would probably not be the best replacement all the time. The PR for that is in https://github.com/flutter/flutter/pull/166099

ash2moon replied:
@reidbaker this PR can be reviewed now since the parent PR was merged

hannah-hyj reviewed (approved):
LGTM with a nit comment

reidbaker reviewed (changes requested):
Please also update the pr description with the current state of the pr. (forked behavior instead of a pure revert) 

@ash2moon see https://docs.flutter.dev/ui/adaptive-responsive/capabilities for a longer set of documenation on my thoughts behind why we should abstract the logic around platform checking to a method.

ash2moon commented:
> Please also update the pr description with the current state of the pr. (forked behavior instead of a pure revert)
> 
> @ash2moon see https://docs.flutter.dev/ui/adaptive-responsive/capabilities for a longer set of documenation on my thoughts behind why we should abstract the logic around platform checking to a method.

Made this PR to resolve this issue:
https://github.com/flutter/flutter/pull/166099

I'll update the code in this PR once that one is merged.

ash2moon commented on packages/flutter/test/material/text_field_test.dart:
```diff
@@ -8892,53 +8892,70 @@ void main() {
-    final SemanticsTester semantics = SemanticsTester(tester);
-    final TextEditingController controller = _textEditingController();
-    final Key key = UniqueKey();
+  for (final bool isAndroid in <bool>[true, false]) {
```
See above comment

dkwingsmt commented:
It seems that the PR has some conflicts. Can you take a look? (from triage)

ash2moon commented:
> It seems that the PR has some conflicts. Can you take a look? (from triage)

Resolved

reidbaker committed:
Update semantic_utils.dart to have licence header

reidbaker commented:
Tests appeared to be failing because of a missing licence in the newly added file. I fixed that to help narrow down issues to the ones wrong with the code.

ash2moon committed:
refactor `SemanticsService.isAnnounceSupported` into DefaultSemanticsService for testability

Piinks commented on packages/flutter/lib/src/semantics/semantics_service.dart:
```diff
@@ -67,11 +70,21 @@ abstract final class SemanticsService {
   /// On other platforms, this will return true.
-  static bool isAnnounceSupported() {
+  @override
+  bool isAnnounceSupported() {
```
@justinmc @chunhtai was talking with @jmagman and wanted to check out this pattern of approach. 'Announce' is deprecated on Android, so we want to got off of it on Android, but the liveRegion/announce support on various platforms is not consistent, so having this check would help determining the right code path.

chunhtai replied:
I think we should use AccessibilityFeatures to avoid having this kind of setting in random places

chunhtai commented on packages/flutter/lib/src/semantics/semantics_service.dart:
```diff
@@ -67,11 +70,21 @@ abstract final class SemanticsService {
   }
 
+  /// Checks if announce is supported on the given platform.
+  bool isAnnounceSupported();
```
we should surface this through AccessibilityFeature
https://api.flutter.dev/flutter/dart-ui/AccessibilityFeatures-class.html

reidbaker replied:
in this pr or a different pr?

Piinks replied:
cc @chunhtai 
I take Chun-Heng's feedback to mean in this PR, as in, we should not add this here, and instead incorporate it into AccessibilityFeature for consistent use across the framework. I'm grabbing some time for folks who know or are working in this area to sync up on this.

ash2moon committed:
revert adding `SemanticsService.isAnnounceSupported`

ash2moon committed:
use noAnnounce for checking if announce is supported

chunhtai commented on packages/flutter/lib/src/material/input_decorator.dart:
```diff
@@ -398,6 +398,8 @@ class _HelperErrorState extends State<_HelperError> with SingleTickerProviderSta
     return Semantics(
       container: true,
+      // Cannot use MediaContext here due to its use in [didUpdateWidget].
+      liveRegion: !PlatformDispatcher.instance.accessibilityFeatures.noAnnounce,
```
Consider wrap the semantics with a Builder widget so that you can get stable context to call mediaQuery.of.

That being said, the code is weird around the _helper and _error widget and feels like they should be a boolean instead of widget. can you add a todo here to refactor them and we will no longer worry about using the context here?

ash2moon replied:
Done, also created a todo link to https://github.com/flutter/flutter/issues/168022

chunhtai reviewed (approved):
LGTM, except for one comment

Piinks reviewed (commented):
Looks like this also needs to be defined in `engine/src/flutter/lib/web_ui/lib/src/engine/semantics/semantics.dart` to resolve the failing web check.
I searched for the preceding feature `onOffSwitchLabels` to see where else the `noAnnounce` might need to be.

Good practice to keep in mind for next time (not here), we would usually try to split these changes up so if one caused an issue, we did not need to revert the whole bunch.
For example here, we'd probably want to revert https://github.com/flutter/flutter/pull/166099 separately, then add the new AccessibilityFeature (maybe + the pipe through to framework semantics and MediaQuery - @chunhtai can speak more to how we usually land those), and then update individual widgets with the feature each in separate changes.

I am _not_ going to ask you to do that here though. 🙂  I was super unclear in my response to @reidbaker's question, that's on me. There is a lot of hard work here. Thank you for that and your patience!

ash2moon committed:
use noAnnounce for checking if announce is supported

cbracken commented on engine/src/flutter/shell/platform/embedder/embedder.h:
```diff
@@ -105,6 +105,8 @@ typedef enum {
   /// Request to show on/off labels inside switches.
   kFlutterAccessibilityFeatureOnOffSwitchLabels = 1 << 6,
+  /// Indicate the platform does not support announcements.
+  kFlutterAccessibilityFeatureNoAnnounce = 1 << 7,
```
embedder API codeowners review: this is safe and will not break ABI compatibility.

matanlurey commented on engine/src/flutter/lib/ui/window.dart:
```diff
@@ -968,6 +969,18 @@ class AccessibilityFeatures {
+  /// announcements are generally supported without discouragement.
+  ///
+  /// Use this flag to conditionally avoid making announcements on Android.
+  bool get noAnnounce => _kNoAnnounceIndex & _index != 0;
```
This is a relatively minor issue, and if you've already discussed this with others more knowledgable of a11y work feel free to disregard - but the negative qualifier _no_ is sort of confusing (to me), at least after reading the description:

> ```dart
>   /// Returns `false` on all other platforms (iOS, web, desktop) where such
>   /// announcements are generally supported without discouragement.
> ```

I'd expect to see something like:

```dart
 /// Whether accessibility announcements (like [SemanticsService.announce])
 /// are encouraged on the current platform.
 ///
 /// Returns `false` on Android, where platform announcements are deprecated
 /// by the underlying platform.
 ///
 /// Returns `true` on all other platforms (iOS, web, desktop) where such
 /// announcements are generally supported without discouragement.
 ///
 /// Use this flag to conditionally make announcements.
bool get announce => _kAnnounceIndex & _index != 0;
```

chunhtai replied:
The issue with positive notion is that for most platform it will be true. This may break exisiting embedder because this is a bit map and default value is 0.

reidbaker replied:
We could rephrase it to be positive for example preferLiveRegion. That said I think we are not actually preferring live region but instead discouraging announce.

matanlurey replied:
I mean a cheap option is just `bool get announce => _kAnnounceIndex & _index == 0`.

Anyway, just trying to get ahead of understanding:
```dart
if (!noAnnounce) { /* ... */ }
```

ash2moon replied:
How about we use `_kNoAnnounceIndex` for the embedder and `announce` for the API?

That way we can avoid confusion?

matanlurey replied:
I care a lot less about the implementation detail, so SGTM

ash2moon committed:
flip noAnnounce to announce

ash2moon committed:
flip noAnnounce to announce

ash2moon committed:
Merge branch 'master' of https://github.com/flutter/flutter into feat/use-liveregion-form

ash2moon committed:
Merge branch 'master' of https://github.com/flutter/flutter into feat/use-liveregion-form

ash2moon committed:
doc update

gmackall commented on packages/flutter/lib/src/semantics/semantics_service.dart:
```diff
@@ -33,8 +33,8 @@ abstract final class SemanticsService {
   ///
-  /// Not all platforms support announcements. Check to see if
-  /// [isAnnounceSupported] before calling this method.
+  /// Not all platforms support announcements. CHeck to see if it is supported using
```
```suggestion
  /// Not all platforms support announcements. Check to see if it is supported using
```

ash2moon committed:
doc update

ash2moon committed:
doc update

ash2moon committed:
Merge branch 'master' of https://github.com/flutter/flutter into feat/use-liveregion-form

ash2moon commented:
Just did my final round of testing on two real devices and made sure the media queries and announcement forking logic works well.

ash2moon committed:
fix flaky test (MediaQueryData.fromView is sane) announce feature changes by platform

ash2moon committed:
fix flaky test (MediaQueryData.fromView is sane) announce feature changes by platform

ash2moon committed:
fix flaky test (MediaQueryData.fromView is sane) announce feature changes by platform

ash2moon committed:
fix flaky test (MediaQueryData.fromView is sane) announce feature changes by platform

ash2moon committed:
fix flaky test (MediaQueryData.fromView is sane) announce feature changes by platform

ash2moon committed:
fix flaky test (MediaQueryData.fromView is sane) announce feature changes by platform

ash2moon committed:
Merge branch 'master' into feat/use-liveregion-form

ash2moon committed:
fix flaky test (MediaQueryData.fromView is sane) announce feature changes by platform

ash2moon committed:
fix flaky test (MediaQueryData.fromView is sane) announce feature changes by platform

matanlurey commented:
reason for revert: Breaks google client tests
